ffi/src/ffi_tracing.rs

```diff
@@ -0,0 +1,624 @@
+/// verbose Levels are "greater than" less verbose ones. So Level::ERROR is the lowest, and
+/// Level::TRACE the highest.
+///
+/// Log lines passed to the callback will already have a newline at the end.
+///
+/// Note that setting up such a call back can only be done ONCE. Calling any of
+/// `enable_event_tracing`, `enable_log_line_tracing`, or `enable_formatted_log_line_tracing` more
+/// than once is a no-op.
+///
+/// Returns `true` if the callback was setup successfully, false on failure (i.e. if called a second
+/// time)
+///
+/// Log line based tracing is simple for an engine as it can just log the passed string, but does
+/// not provide flexibility for an engine to format events. If the engine wants to use a specific
+/// format for events it should call [`enable_event_tracing`] instead of this function.
+///
+/// # Safety
+/// Caller must pass a valid function pointer for the callback
+#[no_mangle]
+pub unsafe extern "C" fn enable_log_line_tracing(
+    callback: TracingLogLineFn,
+    max_level: Level,
+) -> bool {
+    setup_log_line_subscriber(
+        callback,
+        max_level,
+        LogLineFormat::FULL,
+        true, /* ansi color on */
+        true, /* time included */
+        true, /* level included */
+        true, /* target included */
+    )
+    .is_ok()
+}
+
+/// Enable getting called back with log lines in the kernel. This variant allows specifying
+/// formatting options for the log lines. See [`enable_log_line_tracing`] for general info on
+/// getting called back for log lines.
+///
+/// Note that setting up such a call back can only be done ONCE. Calling any of
+/// `enable_event_tracing`, `enable_log_line_tracing`, or `enable_formatted_log_line_tracing` more
+/// than once is a no-op.
+///
+/// Returns `true` if the callback was setup successfully, false on failure (i.e. if called a second
+/// time)
+///
+/// Options that can be set:
+/// - `format`: see [`LogLineFormat`]
+/// - `ansi`: should the formatter use ansi escapes for color
+/// - `with_time`: should the formatter include a timestamp in the log message
+/// - `with_level`: should the formatter include the level in the log message
+/// - `with_target`: should the formatter include what part of the system the event occurred
+///
+/// # Safety
+/// Caller must pass a valid function pointer for the callback
+#[no_mangle]
+pub unsafe extern "C" fn enable_formatted_log_line_tracing(
+    callback: TracingLogLineFn,
+    max_level: Level,
+    format: LogLineFormat,
+    ansi: bool,
+    with_time: bool,
+    with_level: bool,
+    with_target: bool,
+) -> bool {
+    setup_log_line_subscriber(
+        callback,
+        max_level,
+        format,
+        ansi,
+        with_time,
+        with_level,
+        with_target,
+    )
+    .is_ok()
+}
+
+// utility code below for setting up the tracing subscriber for events
+
+pub fn set_global_default(dispatch: tracing_core::Dispatch) -> DeltaResult<()> {
```

Comment:
is this supposed to be `pub`? is it just if consumers want to do something different than `enable_event_tracing`?

Reply:
nope, good catch, removed